src/Piv.js

```diff
@@ -0,0 +1,846 @@
+    const [nowMK, setnowMK] = useState("");
+    const [newMK, setNewMK] = useState("");
+    const [MKValid, setMKValid] = useState(false);  //set to false after test
+    const [MKAuthenDialogOpen,setMKAuthenDialogOpen] = useState(false);
+    const [importCrtDialogOpen, setImportCrtDialogOpen] = useState(false);
+    const [creatCrtDialogOpen, setCreatCrtDialogOpen] = useState(false);
+    const [importKeyDialogOpen, setImportKeyDialogOpen] = useState(false);
+    const [algo, setAlgo] = useState('ECC-P-256');
+    const [country, setCountry] = useState('CN');
+    const [province, setProvince] = useState('Beijing');
+    const [locality, setLocality] = useState('Chiyoda');
+    const [organization, setOrganization] = useState('example');
+    const [organizationUnit, setOrganizationUnit] = useState('Research');
+    const [common, setCommon] = useState('example.com');
+    const [keyPem, setKeyPem] = useState('')
+    const [ownCrt, setOwnCrt] = useState(false)
+    const [keyAlgo, setKeyAlgo] = useState('RSA')
+    const [stopRefresh, SetStopRefresh] = useState(false)
+    let slotPosition = '9a';
+    switch(tabValue) {
+        case 0:
+            slotPosition = '9a';
+            break;
+        case 1:
+            slotPosition = '9c';
+            break;
+        case 2:
+            slotPosition = '9d';
+            break;
+        case 3:
+            slotPosition = '9e';
+            break;
+        default:
+            slotPosition = '9a';
+            break;
+    }
+    const slotMap = {'9a':'5fc105', '9c':'5fc101', '9d':'5fc10a', '9e':'5fc10b'}
+
+    const selectPivApplet = useCallback(async () => {
+        if (device === null) {
+            if (!await dispatch(connect())) {
+                throw new Error('Cannot connect to Canokey');
+            }
+        }
+        let res = await dispatch(transceive("00A4040009A00000030800001000"));
+        if (!res.endsWith("9000")) {
+            throw new Error('Selecting piv applet failed');
+        }
+    }, [device, dispatch]);
+
+    const handleTabChange = (event, newValue) => {
+        setTabValue(newValue);
+    };
+    const onImportKey = useCallback(() => {
+        setImportKeyDialogOpen(true);
+    }, []);
+
+
+
+    const doImportKey = useCallback(async()=>{
+        setImportKeyDialogOpen(false);
+        SetStopRefresh(true);
+        try{
+            if(keyAlgo==='RSA'){
+                const keyObj = new keyutil.Key('pem',keyPem);
+                let jwk = await keyObj.export('jwk');
+
+                let dataString = ''
+                const rsaP = base64url.toBuffer(jwk['p']).toString('hex');
+                dataString = dataString + '018180' + rsaP
+
+                const rsaQ = base64url.toBuffer(jwk['q']).toString('hex');
+                dataString = dataString + '028180' + rsaQ
+
+
+                const rsaDp = base64url.toBuffer(jwk['dp']).toString('hex');
+                dataString = dataString + '038180' + rsaDp

```

Comment:
The lengths of dP, dQ, and Qinv may be less than 128 bytes.

Reply:
I have revised the code